### <img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 3: Web APIs & NLP

Part 3 of 3

**Conclusion and Recommendations**  

In [1]:
import pandas as pd

In [2]:
# Reading the data.
wrongpred = pd.read_csv('../datasets/wrongpred.csv')
print(wrongpred.shape)
wrongpred.head()

(242, 3)


Unnamed: 0  subreddit                                              title
0        3491          0            elon musk has a public mental breakdown
1         211          1                                    the bitter vend
2        3409          0                 at least they wont get that g flu 
3        3836          0  somehow adolf hitler has received a covid vacc...
4        1146          1    home depot releases new bluetooth cordless hose

In [3]:
# Dropping the Unnamed column.
wrongpred.drop(['Unnamed: 0'], axis=1, inplace=True)

In [4]:
# Using this so I can see the complete text in each row.
import warnings
warnings.filterwarnings('ignore')
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)
pd.set_option('display.max_colwidth', None)

In [5]:
# Have a look at the wrong predictions.
wrongpred.rename(columns={"subreddit": "original subreddit"}, inplace=True)
wrongpred

original subreddit  \
0                     0   
1                     1   
2                     0   
3                     0   
4                     1   
5                     1   
6                     1   
7                     1   
8                     0   
9                     1   
10                    0   
11                    1   
12                    0   
13                    0   
14                    0   
15                    1   
16                    1   
17                    1   
18                    0   
19                    0   
20                    1   
21                    1   
22                    0   
23                    0   
24                    0   
25                    1   
26                    1   
27                    1   
28                    1   
29                    0   
30                    1   
31                    0   
32                    0   
33                    1   
34                    1   
35                    0   
36                    1   
37                    1   
38                    0   
39                    1   
40                    0   
41                    1   
42                    0   
43                    0   
44                    1   
45                    1   
46                    0   
47                    1   
48                    1   
49                    0   
50                    1   
51                    1   
52                    0   
53                    1   
54                    1   
55                    0   
56                    0   
57                    1   
58                    1   
59                    0   
60                    1   
61                    1   
62                    1   
63                    1   
64                    1   
65                    0   
66                    1   
67                    0   
68                    0   
69                    1   
70                    0   
71                    0   
72                    1   
73                    0   
74                    0   
75                    0   
76                    0   
77                    1   
78                    0   
79                    1   
80                    0   
81                    1   
82                    0   
83                    1   
84                    1   
85                    1   
86                    1   
87                    0   
88                    1   
89                    0   
90                    0   
91                    1   
92                    1   
93                    0   
94                    1   
95                    0   
96                    0   
97                    0   
98                    1   
99                    0   
100                   1   
101                   0   
102                   0   
103                   0   
104                   1   
105                   1   
106                   0   
107                   1   
108                   1   
109                   0   
110                   1   
111                   0   
112                   1   
113                   1   
114                   1   
115                   0   
116                   0   
117                   1   
118                   0   
119                   0   
120                   1   
121                   1   
122                   1   
123                   0   
124                   0   
125                   0   
126                   1   
127                   1   
128                   1   
129                   1   
130                   0   
131                   1   
132                   0   
133                   1   
134                   1   
135                   1   
136                   0   
137                   0   
138                   1   
139                   1   
140                   0   
141                   1   
142                   1   
143                   1   
144                   1   
145                   1   
146                   0   
147      

As an overview, I scan through the incorrect predictions and pick out a few to analyse below. 

Originally from TheOnion '1' but predicted as from NotTheOnion '0': 

In [7]:
# Picking out a few of the wrong predictions. 
wrongpred.iloc[[16, 27, 53]]

original subreddit  \
16                   1   
27                   1   
53                   1   

                                                                                                   title  
16  bat wielding jim jordan bursts through capitol window demanding to be allowed onto january committee  
27                                       adele single handedly gets youtube to remove the dislike button  
53                                                        pizza delivery driver s sixth grandmother dies

These are actually exaggerated news but were likely predicted as genuine articles because of the words used being rather typical of day to day reporting styles. 

Originally from NotTheOnion '0' but predicted as from TheOnion '1':

In [8]:
wrongpred.iloc[[19, 49, 65]]

original subreddit  \
19                   0   
49                   0   
65                   0   

                                                                                          title  
19                      walking from your bed to home office now counts as commuting in germany  
49  swedish company epicenter develops microchip implant that can be used as a vaccine passport  
65                                      slithering snake burglar steals thousands from ca salon

These are real news but likely got misclassified because of the absurdity within the context. 

As a conclusion, I would say that my model is sufficiently good since it can correctly classify 76% of the posts, which means almost 80% of lost data can be restored accurately. Although I'm rather contented with the performance of my SVC model, I do wonder if implementing a trigram or beyond approach would help me reduce the number of wrong predictions. Having said that, I did run GridSearch which picked out the best ngram_range as \[1,2\] instead of \[1,3\] so I shan't pick on it.

Perhaps the posts from these two subreddit threads are to a certain extent rather similar in terms of their topics and vocabulary used hence my accuracy scores never quite hit the mark of 0.80. 

Considering the accuracy score improved from 0.6+ to 0.7+ by me scrapping an extra 1000 posts per subreddit, I do think my model can do even better if I scrapped more posts. However, with limited time and processing power, the engineer in question should be happy to implement this model in sorting out the posts. On another note, I do believe that if I had more columns of data besides just the title text, my model's performance may be enhanced even further. In the long run, machine learning models are in fact very much feasible for automated moderation in a bit to lighten the human moderator's workload. Having said that, I must also state a caveat that no matter what model we have, such a classification is still somewhat of a blackbox because we cannot say for certain that specific words determine the source of the subreddit since it really depends on the context of the word use. 
